Post-migration tool for de-duplicating `IPAddresses` (#3986)
nautobot/nautobot: nautobot/docs/user-guides/ip-address-merge-tool.md

!!! important
    Possible reasons to maintain duplicate `IP Addresses` can be the following:
    1. You have use cases for duplicate `IP Addresses` with different `nat_inside` attributes.
    2. Duplicate `IP Addresses` are `primary ips` on different `Devices`.

## How to access this tool

To use this tool:
1. Go to the `IPAM` tab on the navigation menu and click on `IP Addresses`. This will take you to the list of `IP Addresses` exist in the database.
2. On the top right of the page, you will find a group of buttons representing different available actions and the first one should be `Find and Merge Duplicate IPs`.
3. Click on that button to access the tool.

![Homepage Navigation](./images/ip-address-merge-tool/navigation.png)

![Merge Button](./images/ip-address-merge-tool/ip_merge_button.png)

> ## How to use this tool

Reviewer: Preceding this section I think it would be good to have a simple "How it works" section that explains the basics covered across some of the lower sections. Specifically we should call out:

- The `IPAddress` found in the chosen `Namespace` you select will be the one used as the base object for the others to be merged into
- If you deselect all other duplicates and attempt to merge, nothing will happen (the operation is a no-op) and the object will be skipped over as if you had clicked the "Skip and Go to the Next Duplicates" button.
- All `Interface` assignments will be automatically updated to point to the merged `IPAddress` object
- All `VMInterface` assignments will be automatically updated to point to the merged IPAddress object
- All `Service` assignments will be automatically updated to point to the merged `IPAddress` object
- All duplicate `IPAddress` objects will be deleted after being merged, leaving only the merged `IPAddress` object in the `Namespace` you selected

And any others?

> Clicking on the `Find and Merge Duplicate IPs` button will automatically query your database for duplicate `IP Addresses` and group them by their respective `host` values. The tool will present the duplicate `IP Addresses` in order from lowest to highest `host` values.

Reviewer: This is actually sorted by highest to lowest `mask_length`, right? Since the `host` value will be the same for each?

Once the duplicate `IP Addresses` are found, the tool will put them in a table ordered by their `mask_length` and the table presents all their selectable attributes. Select the desired attribute values for the eventually merged `IP Address` and click on the `Merge and Go to the Next Duplicates` button to collapse/delete all the `IP Addresses` presented into a new `IP Address` with all the attributes you selected. This will also take you to a page of the next available duplicate `IP Addresses`.

![Merging All Duplicate IPs](./images/ip-address-merge-tool/merge_button.png)

### Merging Only Some Duplicate IP Addresses Presented with Desired Attributes

If you want to keep some of the `IP Addresses` around and merge the others, you can de-select the checkboxes in the first column correpsonding to the `IP Addresses` you do not wish to merge, select the desired attributes and click on the `Merge and Go to the Next Duplicates` button. This operation will only collapse/delete the `IP Addresses` that have their checkboxes selected in the first column.

![Merging Some Duplicate IPs](./images/ip-address-merge-tool/unselect_ips.png)

!!! note

>     Unchecking the checkbox of the corresponding `IP Address` means that the attributes of the `IP Addresses` will not be available for selection. Moreover, If there are only one or less checkboxes checked, clicking on the `Merge and Go to the Next Duplicates` button will be a no-op.

Reviewer: ```suggestion
    Unchecking the checkbox of the corresponding `IP Address` means that the attributes of the `IP Addresses` will not be available for selection. Moreover, If there are only one or less checkboxes checked, clicking on the `Merge and Go to the Next Duplicates` button will cause the IP address to be skipped and not merged.
```

# Duplicate IP Address Merge Post Migration Tool

> Upgrading from Nautobot v1.x to v2.0 can leave some duplicate `IPAddress` objects in the existing database. This tool is designated to help users to reduce unnecessary duplicate `IPAddress` objects by merging them into a single IP Address with all the desired attributes.

Reviewer 2: ```suggestion
In order to satisfy new uniqueness constraints, upgrading from Nautobot v1.x to v2.0 can create duplicate `IPAddress` objects in the existing database. This tool is designated to help users to reduce unnecessary duplicate `IPAddress` objects by merging them into a single IP Address with all the desired attributes.
```

## When to use this tool

> After upgrading to Nautobot v2.0 and running the data migrations necessary, some unnecessarily duplicate `IPAddress` objects might exist in your database. We define duplicate `IPAddress` objects as those which have the same `host` attribute but exist in different `namespaces`. If you have no use case to keep those duplicate `IPAddress` objects around, we recommend you to use this tool to de-duplicate those `IPAddress` objects and keep your database clean and manageable. But if you do have reasons to maintain duplicate `IPAddress` objects, this tool is not for you.

Reviewer 2: ```suggestion
After upgrading to Nautobot v2.0 and running the data migrations necessary, duplicate `IPAddress` objects might exist in your database. We define duplicate `IPAddress` objects as those which have the same `host` attribute but exist in different `namespaces`. If you have no use case to keep those duplicate `IPAddress` objects around, we recommend you to use this tool to de-duplicate those `IPAddress` objects and keep your database clean and manageable. But if you do have reasons to maintain duplicate `IPAddress` objects, this tool is not for you.
```

The duplicates were indeed necessary, but may not be desired 😃

!!! important
    One of the possible reasons to maintain duplicate `IPAddress` objects can be the following:

>     1. You have use cases for duplicate `IPAddress` objects with different `nat_inside` attributes.

Reviewer 2: ```suggestion
    1. You have use cases for duplicate `IPAddress` objects with different `nat_inside` attributes. Only one `nat_inside` IP address can be assigned per object.
```

## How to access this tool

To use this tool:

> 1. Go to the `IPAM` tab on the navigation menu and click on `IP Addresses`. This will take you to the list of `IPAddress` objects exist in the database.

Reviewer 2: ```suggestion
1. Go to the `IPAM` tab on the navigation menu and click on `IP Addresses`. This will take you to the list view of all `IPAddress` objects that exist in the database.
```

2. On the top right of the page, you will find a group of buttons representing different available actions and the first one should be `Find and Merge Duplicate IPs`.
3. Click on that button to access the tool.

![Homepage Navigation](./images/ip-address-merge-tool/navigation.png)

![Merge Button](./images/ip-address-merge-tool/ip_merge_button.png)

## How it works

> 1. All the selected `IPAddress` objects will be deleted and a new `IPAddress` objects with all the desire attributes will be created.

Reviewer 2: ```suggestion
1. All of the selected `IPAddress` objects will be deleted and a new `IPAddress` object with all of the desired attributes will be created.
```

> 2. If you de-select all other duplicates and attempt to merge, nothing will happen (the operation is a no-op) and the object will be skipped over as if you had clicked the `Skip and Go to the Next Duplicate` button.

Reviewer 3: Maybe swap the order of these two list items? It seems weird to me that "step 1" is "Everything will be deleted" and step 2 is "Nothing will happen"

> 3. All `Interface` assignments of the deleted `IPAddress` objects will be automatically updated to point to the newly created `IPAddress` object.
> 4. All `VMInterface` assignments of the deleted `IPAddress` objects will be automatically updated to point to the newly created `IPAddress` object.
> 5. The newly created `IPAddress` object will be added to `IPAddress` assignments of related `Service` objects.
> 6. `primary_ip4/primary_ip6` of `Devices` that are pointing to the deleted `IPAddress` objects will be automatically updated to point to the created `IPAddress` object.

Reviewer 2: ```suggestion
3. All `Interface` assignments of the deleted `IPAddress` objects will be automatically updated to reference the newly created `IPAddress` object.
4. All `VMInterface` assignments of the deleted `IPAddress` objects will be automatically updated to reference the newly created `IPAddress` object.
5. The newly created `IPAddress` object will be added to `IPAddress` assignments of related `Service` objects.
6. `primary_ip4/primary_ip6` of `Devices` that are referencing the deleted `IPAddress` objects will be automatically updated to reference the newly created `IPAddress` object.
```

## How to use this tool

Clicking on the `Find and Merge Duplicate IPs` button will automatically query your database for duplicate `IPAddress` objects and group them by their respective `host` values. The tool will present the duplicate `IPAddress` objects in order from lowest to highest `mask_length` values.

![IP Address Merge View](./images/ip-address-merge-tool/merge_view.png)

> ### Merging All Duplicate IP Addresses Presented with Desired Attributes

Reviewer 3: Let's make these headings a bit more succinct? Plus be consistent at least within this page about sentence case vs title case?

```suggestion
### Merging all presented IP addresses
```

> Once the duplicate `IPAddress` objects are found, the tool will put them in a table ordered by their `mask_length` and the table presents all their editable attributes. Select the desired attribute values for the eventually merged `IPAddress` and click on the `Merge and Go to the Next Duplicate` button to collapse/delete all the `IPAddress` objects presented into a new `IPAddress` with all the attributes you selected. This will also take you to a page of the next available duplicate `IPAddress` objects with a different `host` value.

Reviewer 3: ```suggestion
When a set of potentially duplicate `IPAddress` objects are found, the tool will present them and all of their editable attributes as a table. Select the desired value for each attribute and click on the `Merge and Go to the Next Duplicate` button. Nautobot will combine all of the selected `IPAddress` objects into a new `IPAddress` with the attributes you selected.

If more duplicate addresses are present, with a different `host` value, you will be presented with a new table showing the next set of duplicates, and can repeat this process as needed.
```

![Merging All Duplicate IPs](./images/ip-address-merge-tool/merge_button.png)

> ### Merging Only Some Duplicate IP Addresses Presented with Desired Attributes

Reviewer 3: ```suggestion
### Merging a subset of presented IP addresses
```

> If you want to keep some of the `IPAddress` objects around and merge the others, you can de-select the checkboxes in the first column correpsonding to the `IPAddress` objects you do not wish to merge, select the desired attributes and click on the `Merge and Go to the Next Duplicate` button. This operation will only collapse/delete the `IPAddress` objects that have their checkboxes selected in the first column.

Reviewer 3: ```suggestion
If you want to keep some of the `IPAddress` objects as distinct records but merge others, you can uncheck the check-boxes in the first column corresponding to the objects you do not wish to merge. From the remaining checked records, select the desired attributes and click on the `Merge and Go to the Next Duplicate` button. This operation will only combine the `IPAddress` objects that you have checked.
```

![Merging Some Duplicate IPs](./images/ip-address-merge-tool/unselect_ips.png)

!!! note

>     Unchecking the checkbox of the corresponding `IP Address` means that the attributes of the `IP Addresses` will not be available for selection. Moreover, If there are only one or less checkboxes checked, clicking on the `Merge and Go to the Next Duplicates` button will cause the IP address to be skipped and not merged.

Reviewer 2: ```suggestion
    Unchecking the checkbox of the corresponding `IP Address` means that the attributes of the `IP Addresses` will not be available for selection. Moreover, If there aren't at least two checkboxes checked, clicking on the `Merge and Go to the Next Duplicates` button will cause the IP address to be skipped and not be merged.
```

>     Unchecking the checkbox of the corresponding `IP Address` means that the attributes of the `IP Addresses` will not be available for selection. Moreover, If there are not at least two checkboxes checked, clicking on the `Merge and Go to the Next Duplicate` button will cause the IP address to be skipped and not be merged.

Reviewer 3: ```suggestion
    Unchecking a given `IP Address` prevents its attribute values from being selected for the merge. Moreover, if there are not at least two boxes checked, there is no merging to perform, and so clicking on the `Merge and Go to the Next Duplicate` button will proceed to the next set of duplicate IP addresses without making any changes to the current set.
```
